api/v1beta1/cluster_types.go

```diff
@@ -184,6 +196,32 @@ type MachineDeploymentTopology struct {
 	Variables *MachineDeploymentVariables `json:"variables,omitempty"`
 }
 
+// MachineHealthCheckTopology defines a MachineHealthCheck for a group of machines.
+// MachineHealthCheckTopology can be used to enable or disable the creation of a
+// MachineHealthCheck for the target machines.
+// If specified, it overrides the values provided in the ClusterClass.
+type MachineHealthCheckTopology struct {
+	// Enable controls if a MachineHealthCheck should be creates for the target machines.
+	// If false:
+	//		No MachineHealthCheck will be created.
+	// If not set:
+	//		A MachineHealthCheck will be created if it is defined here or in the associated ClusterClass.
+	//		If not MachineHealthCheck if defined then none will be created.
+	// If true:
+	//		A MachineHealthCheck is guaranteed to be created. Cluster validation will block if `enable` is
+	//		true and no MachineHealthCheck definition is available.
+	// +optional
+	Enable *bool `json:"enable,omitempty"`
+
+	// Spec defines a MachineHealthCheck for a group of machines.
+	// TODO: Rename this to something better.
+	// NOTE: Nesting the MHC definition here so that it can be used to determine if an MHC is defined in the Cluster topology.
+	//       Alternative would be to check each field to see if MHC is defined in topology.
+	//
+	// +optional
+	Spec *MachineHealthCheckClass `json:"spec,omitempty"`
```

Comment:
Had an additional idea. Can you try if you can use the following json tag: 
```
`json:",inline"`
```
(not sure how it works in combination with omitempty or not)

Good case is that we could have the nested level here without having an additional level in YAML/JSON.

But I don't think it will work as I don't think the YAML/JSON parser can map an object only containing enable correctly to the Go object containing a nil spec.

(I'm only aware of that it is used for TypeMeta which is not a pointer)

Reply:
When using 
```go
Spec *MachineHealthCheckClass `json:",inline,omitempty"`
``` 
the `Spec` field is never getting populated and it is always getting treated as nil. The `Enable` field is getting populated though. 

For this to unmarshall properlly (just yaml unmarshalling logic), the correct yaml would would look like:
```
machineHealthCheck:
  enable: true
  spec: 
    maxUnhealthy: "50%"
    ... 
```
but this has 2 problems: 
* It brings back the original nesting problem
* api server rejects this with an error complaing that the field "spec" is undefined becasue is it now not part of the CRD definition

If we wish to avoid nesting here then I think we should move the `Enable` field a level above.
This would mean we will add a new filed called `EnableMachineHealthCheck` to the `ControlPlaneTopology` and `MachineDeploymentTopology` types.

Reply:
Thx for testing this!

> If we wish to avoid nesting here then I think we should move the Enable field a level above.
Agree